File: src/main/scala/s/strategy/WinningStrategy.scala
/* (C) 2022 */
package s.strategy

import java.security.SecureRandom
import java.util.Optional

case class WinningStrategy(
    won: Boolean,
    prevHand: Optional[Hand]
) extends Strategy {
  val random = new SecureRandom()

  def this() {
    this(false, Optional.empty)
  }

  override def nextHand: (Strategy, Hand) = {
    if (!won) {
      val ph = Hand.getHand(random.nextInt(3))
      val st = WinningStrategy(won, Optional.of(ph))
      (st, ph)
    } else (this, prevHand.get)
  }

  override def study(win: Boolean): WinningStrategy =
    WinningStrategy(win, prevHand)
}

File: src/main/java/j/strategy/WinningStrategy.java
package j.strategy;

import io.vavr.Tuple;
import io.vavr.Tuple2;

import java.security.SecureRandom;
import java.util.Optional;

public final class WinningStrategy implements Strategy {
    private final SecureRandom random;
    private final boolean won;
    private final Optional<Hand> prevHand;

    public WinningStrategy(boolean won, Optional<Hand> prevHand) {
        this.random = new SecureRandom();
        this.won = won;
        this.prevHand = prevHand;
    }

    public WinningStrategy() {
        this(false, Optional.empty());
    }

    @Override
    public Tuple2<Strategy, Hand> nextHand() {
        if (!won) {
            var ph = Hand.getHand(random.nextInt(3));
            var st = new WinningStrategy(won, Optional.of(ph));
            return Tuple.of(st, ph);
        }
        return Tuple.of(this, prevHand.get());
    }

    @Override
    public Strategy study(boolean win) {
        return new WinningStrategy(win, prevHand);
    }
}
